Automatically configure mbedtls_platform_zeroize under Windows

andresag01 opened the pull request:
This PR adds functionality to automatically configure Mbed TLS to use the appropriate secure zeroize functions under Windows.

The Windows detection and #define use the code in https://github.com/ARMmbed/mbedtls/pull/1522 as stated in the ChangeLog.

## Status
**READY**

## Requires Backporting
NO

andresag01 commented:
This PR addresses https://github.com/ARMmbed/mbedtls/issues/1584

irwir commented:
Autodetection of C11features should verify that **__STDC_WANT_LIB_EXT1__**  exists and the value is 1 instead of trying to (re)define the macro.
(Re)definition easily creates issues; and the immediate one is failed automatic name validation.

Also, **_MSC_VER** could be replaced with **_WIN32** provided that MinGW never optimizes out _RtlSecureZeroMemory_ calls.

andresag01 commented:
@irwir: Thank you for your review comments.

>Autodetection of C11 features should verify that STDC_WANT_LIB_EXT1 exists and the value is 1 instead of trying to (re)define the macro.

My understanding (from the documentation online) is that the bounds-checking functions are only guaranteed to be available if `__STDC_WANT_LIB_EXT1__` is defined to `1` *before* including `string.h` and after the include the macro `__STDC_LIB_EXT1__` is defined.

> Also, _MSC_VER could be replaced with _WIN32 provided that MinGW never optimizes out RtlSecureZeroMemory calls.

I will check this during testing on Windows.

RonEld commented on include/mbedtls/platform_util.h:
```diff
@@ -25,12 +25,28 @@
+ * string.h and do not define the macro.
+ */
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "mbedtls/config.h"
```
The `mbedtls` namespacing should be removed. See #921

andresag01 commented:
@RonEld: Thanks for the review. What sort of tests are you suggesting? We already have a zeroize test included in all.sh. Also, please note that any tests in windows will check that platform and its very difficult to get a libc that supports memset_s(). I am actually thinking whether we should remove that...

RonEld commented:
@andresag01 Disregard my comment on the tests part, if there are already tests for the zeroize (to be honest I haven't checked),
If we are not able to test `memset_s`, then I don't think we should supply it.

k-stachowiak commented on include/mbedtls/platform_util.h:
```diff
@@ -54,6 +70,32 @@ extern "C" {
 void mbedtls_platform_zeroize( void *buf, size_t len );
+#elif defined(_MSC_VER)
+/* Windows implementation */
+#include <wdm.h>
```
Additionally my distribution of MinGW did not have the RtlSecureZeroMemory function defined, so I think that environment should not be handled in the same branch as MSVC.

andresag01 replied:
Thanks! I will take a look at this.

andresag01 replied:
@k-stachowiak: Could you please add the version of MingW that you used for reference? Thanks!

k-stachowiak replied:
I am not certain if I can give you the Mingw version, however the gcc in my distribution is 6.3.0

irwir commented:
> My understanding (from the documentation online) is that the bounds-checking functions are only guaranteed to be available if __STDC_WANT_LIB_EXT1__ is defined to 1 before including string.h and after the include the macro __STDC_LIB_EXT1__ is defined. 

That is only a part of standard's requirements for a **C11 compiler**.
Forcing definition of the macro means that the fact **'this is a C11 compiler'** was already known or assumed.
That hardly qualifies as an _autodetection_.

> I will check this during testing on Windows.

MSDN explicitly specified that `RtlSecureZeroMemory` should never be optimized out.
Testing cannot guarantee that, only compiler makers (gcc/MinGW) could. I know they are trying to follow MS guidelines, but quick search in the net did not bring a reliable proof.

irwir commented:
Would it be necessary to add two files into the "production" library?
The name `platform_util` implies general purposes, but there is already `platform.c` module, which is small enough to host `zeroize` functions as well.

andresag01 commented:
@irwir: Thanks for reviewing. Please find my reply to each of your comments below.

> At least some macros in platform_util.h could be made paremeterless for the existing implementations. After removing ( buf, len ) for Windows platform, the solution still builds successfully.

Indeed they could, but (as you pointed out) there are some functions (such as memset_s()) that need the parenthesis because their declaration does not match that of mbedtls_platform_zeroize(). Given this, I prefer to be consistent and put the parenthesis in all the defines in the library.

> Would it be necessary to add two files into the "production" library?
The name platform_util implies general purposes, but there is already platform.c module, which is small enough to host zeroize functions as well.

Strictly speaking, it is not necessary. However, we have decided to split the platform module functionally across multiple source files to keep it tidy and accessible. I think there are three main files: platform, platform_time and platform_util. At the moment, platform_util only contains one function (mbedtls_platform_zeroize()), but I expect this to grow in the future as needed.

> Header files in platform_util.c, lines 31-32, were already included in platform_util.h and thus should be removed.

Thanks for pointing this out, I will remove them. However, having said that, it is generally my preference to put all includes that a file needs instead of relying on intermediate files to include them. I think that makes code more explicit and in any case it does not make a difference because all headers are guarded with `#ifdef` directives.

> RtlSecureZeroMemory should be present since gcc v 4.8.2

I got a hold of a Windows system and tried this for myself on an installation of MingW with gcc v.5.3.x. The only change to the patch is changing [this line](https://github.com/ARMmbed/mbedtls/pull/1644/files#diff-7d2c8f790aa5998c269656d95de4851cR42) to use `_WIN32` instead of `_MSC_VER`. As you pointed out, RtlSecureZeroMemory was present and the library compiled. However, the compiler kept issuing the following warning regardless of the includes I put in:
```
warning: implicit declaration of function 'RtlSecureZeroMemory'
```
Do you know which one is the appropriate header to include for MingW? Please note that I am not very familiar with Windows and MingW.

irwir commented:
> there are some functions (such as memset_s()) that need the parenthesis because their declaration does not match that of mbedtls_platform_zeroize().

Macro is platform dependent anyway, so maybe the simpler the better.

> Do you know which one is the appropriate header to include for MingW?

In my installation `RtlSecureZeroMemory` was found in:
```
\MinGW\x86_64-w64-mingw32\include\ddk\wdm.h
\MinGW\x86_64-w64-mingw32\include\winbase.h
\MinGW\x86_64-w64-mingw32\include\winnt.h 
```

irwir commented:
> I think that makes code more explicit and in any case it does not make a difference

For example, PC-lint issues warnings in such cases.

andresag01 commented:
@irwir:
> In my installation RtlSecureZeroMemory was found in:

Thanks for the information. It turned out that there was a problem with my MingW installation. I fixed this and it now builds cleanly with the regular "windows.h" include. Also, I manually looked at the disassembly of the zeroize.c test program to see if the compiler is optimising away the call to RtlSecureZeroMemory() and I got the following fragment of disassembly:
```
...
  402cfd:	e8 c6 fd ff ff       	callq  402ac8 <fclose>
  402d02:	b9 00 04 00 00       	mov    $0x400,%ecx
  402d07:	31 c0                	xor    %eax,%eax
  402d09:	f3 aa                	rep stos %al,%es:(%rdi)
  402d0b:	eb 83                	jmp    402c90 <main+0x30>
...
```
Which I think corresponds to the following fragment of C code:
```
    fclose( fp );
    mbedtls_platform_zeroize( buf, sizeof( buf ) );
```
So it seems that the compiler is not optimising away the zeroize call. I will change the preprocessor directives so that the RtlSecureZeroMemory is also used when compiling with MingW

andresag01 commented:
@RonEld @k-stachowiak: Could you please review once again?

k-stachowiak commented on include/mbedtls/platform_util.h:
```diff
@@ -54,6 +63,18 @@ extern "C" {
+#include <windows.h>
+#define MBEDTLS_PLATFORM_ZEROIZE_MACRO( buf, len )  \
+    RtlSecureZeroMemory( buf, len )
+#endif /* MBEDTLS_PLATFORM_ZEROIZE_ALT || !_MSC_VER */
```
Minor remark: the test we close here checks _WIN32 rather than _MSC_VER

andresag01 replied:
Thanks. Will do that.

irwir commented:
Something still could be improved.
1. There is one macro too many: MBEDTLS_PLATFORM_ZEROIZE_MACRO could be eliminated.
Alternative implementations could be controlled with MBEDTLS_PLATFORM_ZEROIZE_ALT, and implementation itself could be either function- or macro-based.
2. Creating an alternative implementation for Windows (for testing, maybe) is not the same as for other platforms. Maybe it is not very necessary, but treating all platforms in a more uniform way would be preferable.

Changes would rearrange the code substantially.
Perhaps I will need to create a new PR and post a link here.

andresag01 commented:
@irwir: Thank you for your input. Could you please elaborate a bit more on your comment?

> There is one macro too many: MBEDTLS_PLATFORM_ZEROIZE_MACRO could be eliminated.
Alternative implementations could be controlled with MBEDTLS_PLATFORM_ZEROIZE_ALT, and implementation itself could be either function- or macro-based.

How do you propose to eliminate MBEDTLS_PLATFORM_ZEROIZE_MACRO? The idea of having MBEDTLS_PLATFORM_ZEROIZE_MACRO is that it controls the macro-based implementation while MBEDTLS_PLATFORM_ZEROIZE_ALT controls the function-based implementation. So isn't the second part of your comment already implemented by this PR?
Please note that the idea of having two macros to control these functions is this way in the PR as it follows the patter already found in other platform features, such as replacement for standard library functions, hardware acceleration, etc. The idea is to be consistent...

> Creating an alternative implementation for Windows (for testing, maybe) is not the same as for other platforms. Maybe it is not very necessary, but treating all platforms in a more uniform way would be preferable.

How would you propose to treat all platforms in a more uniform way? Are you suggesting that the autodetection mechanism and Windows implementation for mbedtls_platform_zeroize used in this PR is only useful for testing? Please bear in mind that the Windows implementation and autodetection is a change I took from [*your PR*](https://github.com/ARMmbed/mbedtls/pull/1522)...

andresag01 commented:
@irwir: Once more, thank you for your input, but I have my reservations about your patch at #1760. In my opinion, it is not consistent with the way the _ALT macros generally work. Also, its more limiting that the change I have proposed here. Having said that, either change has to be considered for design approval before it is merged, so I will leave the decision to our gatekeepers.

andresag01 commented:
@k-stachowiak: Thanks for reviewing, I have fixed the mismatch in the endif comment.

irwir commented:
> In my opinion, it is not consistent with the way the _ALT macros generally work.
> Also, its more limiting that the change I have proposed here.

Could you please to tell more on these two points for better understanding?

sbutcher-arm commented:
@andresag01 - Can you please rebase this? There are some conflicts.

andresag01 commented:
@sbutcher-arm: I have rebased the changes on top of development and cleaned up the commit history for this PR a little bit. Could you please review for design approval as this adds to the API?

andresag01 commented on include/mbedtls/config.h:
```diff
@@ -3128,6 +3128,17 @@
+ * RtlSecureZeroMemory when compiling for Windows. Refer to platform_utils.h
+ * for more information.
+ */
+//#define MBEDTLS_PLATFORM_ZEROIZE_MACRO( buf, len )    memset( buf, 0, len )
```
It was just an example usage, so I happy to change it.

andresag01 commented:
@k-stachowiak: Thanks for reviewing, I addressed your comment.

daverodgman commented:
This is tracked by #1584. Keeping the issue open, but closing this as this has been inactive for some time & needs work.
